servers/src/main/java/tachyon/master/block/BlockMaster.java

```diff
@@ -0,0 +1,648 @@
+   * be locked first.
+   */
+  private final Map<Long, MasterBlockInfo> mBlocks = new HashMap<Long, MasterBlockInfo>();
+  /**
+   * Keeps track of block which are no longer in tachyon storage. Access must be synchronized on
+   * mBlocks.
+   */
+  private final Set<Long> mLostBlocks = new HashSet<Long>();
+  /** This state must be journaled. */
+  private final BlockIdGenerator mBlockIdGenerator = new BlockIdGenerator();
+
+  /** Worker metadata management. */
+  private final IndexedSet.FieldIndex<MasterWorkerInfo> mIdIndex =
+      new IndexedSet.FieldIndex<MasterWorkerInfo>() {
+        @Override
+        public Object getFieldValue(MasterWorkerInfo o) {
+          return o.getId();
+        }
+      };
+  private final IndexedSet.FieldIndex<MasterWorkerInfo> mAddressIndex =
+      new IndexedSet.FieldIndex<MasterWorkerInfo>() {
+        @Override
+        public Object getFieldValue(MasterWorkerInfo o) {
+          return o.getAddress();
+        }
+      };
+  @SuppressWarnings("unchecked")
+  /**
+   * All worker information. Access must be synchronized on mWorkers. If both block and worker
+   * metadata must be locked, mBlocks must be locked first.
+   */
+  private final IndexedSet<MasterWorkerInfo> mWorkers =
+      new IndexedSet<MasterWorkerInfo>(mIdIndex, mAddressIndex);
+  /**
+   * Keeps track of workers which are no longer in communication with the master. Access must be
+   * synchronized on mWorkers.
+   */
+  private final BlockingQueue<MasterWorkerInfo> mLostWorkers =
+      new LinkedBlockingQueue<MasterWorkerInfo>();
+  /** The service that detects lost worker nodes, and tries to restart the failed workers. */
+  private Future<?> mLostWorkerDetectionService;
+  /** The next worker id to use. This state must be journaled. */
+  private final AtomicLong mNextWorkerId = new AtomicLong(1);
+
+  /**
+   * @param baseDirectory the base journal directory
+   * @return the journal directory for this master
+   */
+  public static String getJournalDirectory(String baseDirectory) {
+    return PathUtils.concatPath(baseDirectory, Constants.BLOCK_MASTER_SERVICE_NAME);
+  }
+
+  public BlockMaster(Journal journal, TachyonConf tachyonConf) {
+    super(journal,
+        Executors.newFixedThreadPool(2, ThreadFactoryUtils.build("block-master-%d", true)));
+    mTachyonConf = tachyonConf;
+  }
+
+  @Override
+  public TProcessor getProcessor() {
+    return new BlockMasterService.Processor<BlockMasterServiceHandler>(
+        new BlockMasterServiceHandler(this));
+  }
+
+  @Override
+  public String getServiceName() {
+    return Constants.BLOCK_MASTER_SERVICE_NAME;
+  }
+
+  @Override
+  public void processJournalCheckpoint(JournalInputStream inputStream) throws IOException {
+    // clear state before processing checkpoint.
+    mBlocks.clear();
+    super.processJournalCheckpoint(inputStream);
+  }
+
+  @Override
+  public void processJournalEntry(JournalEntry entry) throws IOException {
```

Comment:
instead of doing a switch on the type of JournalEntry, it is better to use polymorphism. In particular, different instances of JournalEntry could define a "apply()" method that inputs the BlockMaster state and modify it. 

Feel free to add a TODO for this for me and I will take a stab once this PR is merged.